presto-common/src/main/java/com/facebook/presto/common/zorder/ZOrderIntegers.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common.zorder;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.String.format;
+
+public class ZOrderIntegers
+        implements ZOrder
+{
+    public static final int MAX_INPUT_SIZE = 10;
+
+    // Assume mostSignificantSetBitPositions is 1-based
+    private List<Integer> mostSignificantSetBitPositions;
+
+    public ZOrderIntegers(List<Integer> mostSignificantSetBitPositions)
+    {
+        this.mostSignificantSetBitPositions = mostSignificantSetBitPositions;
+    }
+
+    private void checkInputValidity(List<Integer> input)
+            throws IllegalArgumentException
+    {
+        if (input.size() == 0) {
+            throw new IllegalArgumentException("Input list size should be greater than zero.");
+        }
+
+        if (input.size() > MAX_INPUT_SIZE) {
+            throw new IllegalArgumentException("Our current Z-Ordering implementation does not support more than 10 input numbers.");
+        }
+
+        if (input.size() > mostSignificantSetBitPositions.size()) {
+            throw new IllegalArgumentException(format("a (%d) does not match bit position list size (%d).",
+                    input.size(), mostSignificantSetBitPositions.size()));
+        }
+
+        for (int index = 0; index < input.size(); index++) {
+            if (input.get(index) < 0) {
+                throw new IllegalArgumentException("Our current Z-Ordering implementation does not support negative input numbers.");
+            }
+        }
+    }
+
+    // TODO: make int and long z-address version
+    public byte[] encode(List input)
+            throws IllegalArgumentException
+    {
+        checkInputValidity(input);
+
+        int totalBitLength = 0;
+        int maxBitLength = 0;
+        for (int bitLength : mostSignificantSetBitPositions) {
+            totalBitLength += bitLength;
+            maxBitLength = Math.max(maxBitLength, bitLength);
+        }
+
+        byte[] address = new byte[(totalBitLength + 7) >> 3];
+        int bitIndex = totalBitLength - 1;
+        int bitOffset = bitIndex & 7;
+        // Interweave input bits into address from the most significant bit to preserve data locality
+        for (int inputBitPos = 0; inputBitPos < maxBitLength; inputBitPos++) {
```

Comment:
inputBitPos is a confusing name, since it implies that encoding starts from the lowest bit. Any better name?